libfreerdp/core/aad.c

```diff
@@ -0,0 +1,1148 @@
+
+	if (status_code != 200)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "%s HTTP status code: %li", status_code);
+		free(buffer);
+		return NULL;
+	}
+	*plen = length;
+	return buffer;
+}
+
+static BOOL aad_read_and_extract_token_from_json(rdpAad* aad, BIO* bio)
+{
+	BOOL rc = FALSE;
+	size_t blen = 0;
+	char* buffer = aad_read_response(aad, bio, &blen, "access token");
+	if (!buffer)
+		return FALSE;
+
+	cJSON* json = cJSON_ParseWithLength(buffer, blen);
+	if (!json)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "Failed to parse JSON response");
+		goto fail;
+	}
+
+	if (!json_get_string_alloc(aad->log, json, "access_token", &aad->access_token))
+	{
+		WLog_Print(aad->log, WLOG_ERROR,
+		           "Could not find \"access_token\" property in JSON response");
+		goto fail;
+	}
+
+	rc = TRUE;
+fail:
+	free(buffer);
+	cJSON_free(json);
+	return rc;
+}
+
+static BOOL aad_read_and_extrace_nonce_from_json(rdpAad* aad, BIO* bio)
+{
+	BOOL rc = FALSE;
+	size_t blen = 0;
+	char* buffer = aad_read_response(aad, bio, &blen, "Nonce");
+	if (!buffer)
+		return FALSE;
+
+	/* Extract the nonce from the response */
+	cJSON* json = cJSON_ParseWithLength(buffer, blen);
+	if (!json)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "Failed to parse JSON response");
+		goto fail;
+	}
+
+	if (!json_get_string_alloc(aad->log, json, "Nonce", &aad->nonce))
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "Could not find \"Nonce\" property in JSON response");
+		goto fail;
+	}
+	rc = TRUE;
+fail:
+	free(buffer);
+	cJSON_free(json);
+	return rc;
+}
+
+static BOOL aad_send_token_request(rdpAad* aad, BIO* bio, const char* auth_code)
+{
+	BOOL rc = FALSE;
+
+	char* req_body = NULL;
+	char* req_header = NULL;
+	size_t req_body_len = 0;
+	size_t req_header_len = 0;
+	const int trc = alloc_sprintf(&req_body, &req_body_len, token_http_request_body, auth_code,
+	                              aad->hostname, aad->kid);
+	if (trc < 0)
+		goto fail;
+	const int trh = alloc_sprintf(&req_header, &req_header_len, token_http_request_header, trc);
+	if (trh < 0)
+		goto fail;
+
+	WLog_Print(aad->log, WLOG_DEBUG, "HTTP access token request: %s%s", req_header, req_body);
+
+	if (!aad_logging_bio_write(aad, bio, req_header))
+		goto fail;
+	if (!aad_logging_bio_write(aad, bio, req_body))
+		goto fail;
+	rc = TRUE;
+fail:
+	free(req_body);
+	free(req_header);
+	return rc;
+}
+
+int aad_client_begin(rdpAad* aad)
+{
+	int ret = -1;
+	SSL_CTX* ssl_ctx = NULL;
+	BIO* bio = NULL;
+	char* auth_code = NULL;
+
+	WINPR_ASSERT(aad);
+	WINPR_ASSERT(aad->rdpcontext);
+
+	rdpSettings* settings = aad->rdpcontext->settings;
+	WINPR_ASSERT(settings);
+
+	freerdp* instance = aad->rdpcontext->instance;
+	WINPR_ASSERT(instance);
+
+	/* Get the host part of the hostname */
+	const char* hostname = freerdp_settings_get_string(settings, FreeRDP_ServerHostname);
+	if (!hostname)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "FreeRDP_ServerHostname == NULL");
+		return -1;
+	}
+
+	aad->hostname = _strdup(hostname);
+	if (aad->hostname)
```

Comment:
We hope for it to *not* be null...

Reply:
why? just fail if something works is much more fun :)
damn, that is a nasty one